sstables/mx/partition_reversing_data_source.cc

```diff
@@ -0,0 +1,442 @@
+    }
+    temporary_buffer<char> last_bytes(size_t len) {
+        auto tmp = _cached_read.share(_cached_read.size() - len, len);
+        _cached_read.trim(_cached_read.size() - len);
+        return tmp;
+    }
+
+    column_translation get_column_translation() {
+        return _sst->get_column_translation(_schema, _sst->get_serialization_header(), _sst->features());
+    }
+    void modify_cache(const row_body_skipping_context::tombstone_reversing_info& info) {
+        auto to_idx = [this] (uint64_t offset) {
+            return _cached_read.size() - (_row_end - offset);
+        };
+        _cached_read.get_write()[to_idx(info.kind_offset)] = 7 - (int)info.range_tombstone_kind;
+        if ((int)info.range_tombstone_kind < 2 || (int)info.range_tombstone_kind > 5) {
+            auto before = _cached_read.share(to_idx(info.before_offset), info.after_offset - info.before_offset);
+            auto after = _cached_read.share(to_idx(info.after_offset), _row_end - info.after_offset);
+            std::copy(after.begin(), after.end(), _cached_read.get_write() + to_idx(info.before_offset));
+            std::copy(before.begin(), before.end(), _cached_read.get_write() + to_idx(info.before_offset) + after.size());
+        }
+    }
+public:
+    partition_reversing_data_source_impl(const schema& s,
+            shared_sstable sst,
+            index_reader& ir,
+            uint64_t partition_start,
+            size_t partition_len,
+            reader_permit permit,
+            const io_priority_class& io_priority,
+            tracing::trace_state_ptr trace_state)
+        : _schema(s)
+        , _sst(std::move(sst))
+        , _ir(ir)
+        , _io_priority(io_priority)
+        , _permit(std::move(permit))
+        , _trace_state(std::move(trace_state))
+        , _partition_start(partition_start)
+        , _partition_end(partition_start + partition_len)
+        , _row_start(_partition_end)
+        , _row_end(_partition_end)
+    { }
+
+    partition_reversing_data_source_impl(partition_reversing_data_source_impl&&) noexcept = default;
+    partition_reversing_data_source_impl& operator=(partition_reversing_data_source_impl&&) noexcept = default;
+
+    virtual future<temporary_buffer<char>> get() override {
+        if (!_partition_header_context) {
+            _partition_header_context.emplace(data_stream(_partition_start, _partition_end), _partition_start, _partition_end - _partition_start, _permit);
+            co_await _partition_header_context->consume_input();
+            _clustering_range_start = _partition_header_context->header_end_pos();
+            co_return co_await data_read(_partition_start, _clustering_range_start);
+        }
+        if (_ir.data_file_positions().end && *_ir.data_file_positions().end < _row_end) {
+            // our way of noticing a new clustering range
+            _row_start = *_ir.data_file_positions().end;
+            _state = state::RANGE_END;
+        }
```

Comment:
How does fast_forward_to(position_range) work? It interacts with the index directly, so it works with non-reversed offsets.

Reply:
Did it clarify [here](https://github.com/scylladb/scylla/pull/9171#discussion_r685806342)?
According to https://github.com/scylladb/scylla/blob/4ae6eae00a90ecf7cccf1ec7a2bd45cd06a2344b/flat_mutation_reader_v2.hh#L394-L398
fast_forward_to(position_range) is not supposed to be used on single partition readers, so it does not work with reverse reads right either
We could probably modify it so that it works, but I'm not even sure if the range that we're forwarding to is reversed and I can't really find places where we're using this, even on multiple partition reads (clustering_order_reader_merger::peek_reader looks like a candidate)

Reply:
I think that comment only applies to the `partition_range` variant. 

Btw, we should eventually respect the fwd_mr flag in deciding whether f_f_t() across partitions is supported. The problem is that it's yes by default, so we would lose the single-partition optimization, which benefits from not having to worry about wider partition range. There were patches on the mailing list to fix that but they caused some unit test failures which were not investigated. 

The range passed to reader's f_f_t() should conform to the reader's schema, so reversed. Reversing reader wrapper should intercept this and flip.